File: problems/2.add_two_numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

#include <stdlib.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

#ifndef NULL
#define NULL ((void *)0)
#endif

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2)
{
    struct ListNode *p, *lastp;
    p = l1;
    int carry = 0;
    int val2 = 0;
    while(p) {
        if (l2) {
            val2 = l2->val;
            l2 = l2->next;
        } else {
            val2 = 0;
        }
        p->val = p->val + val2 + carry;
        carry = p->val / 10;
        p->val = p->val % 10;
        lastp = p;
        p = p->next;
    }
    
    while(l2) {
        if (carry) {
            l2->val = l2->val + carry;
            carry = l2->val / 10;
            l2->val = l2->val % 10;
        }
        lastp->next = l2;
        lastp = l2;
        l2 = l2->next;
    }

    if (carry) {
        lastp->next = (struct ListNode *)malloc(sizeof(struct ListNode));
        lastp->next->next = NULL;
        lastp->next->val = carry;
        lastp = lastp->next;
    }
    return l1;
}

File: interview_top/2.add_two_numbers.go
package interviewtop

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{}
	flag := 0
	cursor := head
	for l1 != nil && l2 != nil {
		val := l1.Val + l2.Val + flag
		if val > 9 {
			val = val - 10
			flag = 1
		} else {
			flag = 0
		}
		cursor.Next = &ListNode{
			Val: val,
		}
		cursor, l1, l2 = cursor.Next, l1.Next, l2.Next
	}
	for l1 != nil {
		val := l1.Val + flag
		if val > 9 {
			val = val - 10
			flag = 1
		} else {
			flag = 0
		}
		cursor.Next = &ListNode{
			Val: val,
		}
		cursor, l1 = cursor.Next, l1.Next
	}
	for l2 != nil {
		val := l2.Val + flag
		if val > 9 {
			val = val - 10
			flag = 1
		} else {
			flag = 0
		}
		cursor.Next = &ListNode{
			Val: val,
		}
		cursor, l2 = cursor.Next, l2.Next
	}
	if flag == 1 {
		cursor.Next = &ListNode{
			Val: flag,
		}
	}
	return head.Next
}
